### PR TITLE
Add some extra checks to BinaryReader/Writer buffers

### PR DESCRIPTION
**Pull request by ianhays**
initial work done by @ellismg 

cc: @AlexGhiondea @JeremyKuhne @morganbr @stephentoub 

release/1.1.0 port to come after merge.

resolves https://github.com/dotnet/corefx/issues/12837

### CODE REVIEW

File: src/mscorlib/src/System/IO/BinaryReader.cs [hunk 1]
```diff
@@ -372,10 +372,26 @@ public void Dispose()
...
 
                 Contract.Assert(byteBuffer != null, "expected byteBuffer to be non-null");
-                unsafe {
-                    fixed (byte* pBytes = byteBuffer)
-                    fixed (char* pChars = buffer) {
-                        charsRead = m_decoder.GetChars(pBytes + position, numBytes, pChars + index, charsRemaining, false);
+
+                checked
+                {
+                    if (position < 0 || numBytes < 0 || position + numBytes > byteBuffer.Length)
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/BinaryReader.cs]: Should ```position + numBytes > byteBuffer.Length``` instead be ```position > byteBuffer.Length - numBytes```?

File: src/mscorlib/src/System/IO/BinaryReader.cs [hunk 2]
```diff
@@ -372,10 +372,26 @@ public void Dispose()
...
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(byteCount));
+                    }
+                    if (index < 0 || charsRemaining < 0 || index + charsRemaining > buffer.Length)
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/BinaryReader.cs]: Same question

File: src/mscorlib/src/System/IO/BinaryReader.cs [hunk 3]
```diff
@@ -372,10 +372,26 @@ public void Dispose()
...
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(charsRemaining));
+                    }
+                    unsafe
+                    {
+                        fixed (byte* pBytes = byteBuffer)
+                        {
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/BinaryReader.cs]: Nit: this extra level of braces isn't needed

File: src/mscorlib/src/System/IO/BinaryReader.cs [hunk 4]
```diff
@@ -372,10 +372,26 @@ public void Dispose()
...
+                            fixed (char* pChars = buffer)
+                            {
+                                charsRead = m_decoder.GetChars(pBytes + position, numBytes, pChars + index, charsRemaining, false);
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/BinaryReader.cs]: Nit: I realize it was this way before, but could you use a named argument for the Boolean to help make it clear what ```false``` means?

File: src/mscorlib/src/System/IO/BinaryWriter.cs [hunk 5]
```diff
@@ -383,14 +384,24 @@ public unsafe virtual void Write(String value)
...
-                    fixed(char* pChars = value) {
-                        fixed(byte* pBytes = _largeByteBuffer) {
-                            byteLen = _encoder.GetBytes(pChars + charStart, charCount, pBytes, LargeByteBufferSize, charCount == numLeft);
+
+                    checked
+                    {
+                        if (charStart < 0 || charCount < 0 || charStart + charCount > value.Length)
+                        {
+                            throw new ArgumentOutOfRangeException(nameof(charCount));
+                        }
```
Comment (stephentoub) [inline src/mscorlib/src/System/IO/BinaryWriter.cs]: Same question / comment

**Comment by ianhays** [comment]
Thanks Steve, I pushed an update responding to all of your feedback.

**Pull request merged**